# NS5001

<table>
<tr>
  <td>CheckId</td>
  <td>NS5001</td>
</tr>
<tr>
  <td>Category</td>
  <td>Usage</td>
</tr>
</table>

## Cause

Usage of received-like method in `Received.InOrder` callback.

## Rule description

A violation of this rule occurs when any of the following are used inside a `Received.InOrder` callback:

- `Received()`
- `ReceivedWithAnyArgs()`
- `DidNotReceive()`
- `DidNotReceiveWithAnyArgs()`

Calls within `Received.InOrder` are already checked to ensure they were received in the expected order. Individual received-like assertions should be moved outside the `Received.InOrder` callback.

## How to fix violations

To fix a violation of this rule, remove received-like method calls from `Received.InOrder` callback.

For example:

````c#
// Incorrect:
Received.InOrder(() =>
{
    sub.Received().Baz();
    sub.Received().Bar();
})

// Correct:
Received.InOrder(() =>
{
    sub.Baz();
    sub.Bar();
})
````

Alternatively, move any received-like methods outside of `Received.InOrder` block if ordering is not important:

````c#
// Incorrect:
Received.InOrder(() =>
{
    sub.Baz();
    sub.Zap();
    sub.DidNotReceive().Bar();
})

// Correct:
Received.InOrder(() =>
{
    sub.Baz();
    sub.Zap();
})
sub.DidNotReceive().Bar();
````

## How to suppress violations

This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
The warning can also be suppressed programmatically for an assembly:
````c#
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "NS5001:Received-like method used in Received.InOrder block.", Justification = "Reviewed")]
````

Or for a specific code block:
````c#
#pragma warning disable NS5001 // Received-like method used in Received.InOrder block.
// the code which produces warning
#pragma warning restore NS5001 // Received-like method used in Received.InOrder block.
